src/camera/pipewire/SDL_camera_pipewire.c

```diff
@@ -0,0 +1,1181 @@
+{
+    struct global *g = object;
+    uint32_t i;
+
+    info = g->info = PIPEWIRE_pw_node_info_merge(g->info, info, g->changed == 0);
+    if (info == NULL)
+        return;
+
+    if (info->change_mask & PW_NODE_CHANGE_MASK_PARAMS) {
+        for (i = 0; i < info->n_params; i++) {
+            uint32_t id = info->params[i].id;
+            int res;
+
+            if (info->params[i].user == 0)
+                continue;
+            info->params[i].user = 0;
+
+	    if (id != SPA_PARAM_EnumFormat)
+		    continue;
+
+            param_add(&g->pending_list, info->params[i].seq, id, NULL);
+            if (!(info->params[i].flags & SPA_PARAM_INFO_READ))
+                continue;
+
+            res = pw_node_enum_params((struct pw_node*)g->proxy,
+                        ++info->params[i].seq, id, 0, -1, NULL);
+            if (SPA_RESULT_IS_ASYNC(res))
+                info->params[i].seq = res;
+
+	    g->changed++;
+        }
+    }
+    do_resync();
+}
+
+static void node_event_param(void *object, int seq,
+                uint32_t id, uint32_t index, uint32_t next,
+                const struct spa_pod *param)
+{
+    struct global *g = object;
+    param_add(&g->pending_list, seq, id, param);
+}
+
+static const struct pw_node_events node_events = {
+    .version = PW_VERSION_NODE_EVENTS,
+    .info = node_event_info,
+    .param = node_event_param,
+};
+
+static void node_destroy(struct global *g)
+{
+    if (g->info) {
+        PIPEWIRE_pw_node_info_free(g->info);
+        g->info = NULL;
+    }
+}
+
+
+static const struct global_class node_class = {
+    .type = PW_TYPE_INTERFACE_Node,
+    .version = PW_VERSION_NODE,
+    .events = &node_events,
+    .destroy = node_destroy,
+};
+
+/** proxy */
+static void proxy_removed(void *data)
+{
+    struct global *g = data;
+    PIPEWIRE_pw_proxy_destroy(g->proxy);
+}
+
+static void proxy_destroy(void *data)
+{
+    struct global *g = data;
+    spa_list_remove(&g->link);
+    g->proxy = NULL;
+    if (g->class) {
+        if (g->class->events)
+            spa_hook_remove(&g->object_listener);
+        if (g->class->destroy)
+            g->class->destroy(g);
+    }
+    param_clear(&g->param_list, SPA_ID_INVALID);
+    param_clear(&g->pending_list, SPA_ID_INVALID);
+    free(g->name);
+}
+
+static const struct pw_proxy_events proxy_events = {
+    .version = PW_VERSION_PROXY_EVENTS,
+    .removed = proxy_removed,
+    .destroy = proxy_destroy
+};
+
+// called with thread_loop lock
+static void hotplug_registry_global_callback(void *object, uint32_t id,
+		uint32_t permissions, const char *type, uint32_t version,
+		const struct spa_dict *props)
+{
+    const struct global_class *class = NULL;
+    struct pw_proxy *proxy;
+    const char *str, *name = NULL;
+
+    if (spa_streq(type, PW_TYPE_INTERFACE_Node)) {
+        if (props == NULL)
+            return;
+        if (((str = spa_dict_lookup(props, PW_KEY_MEDIA_CLASS)) == NULL) ||
+            (!spa_streq(str, "Video/Source")))
```

Comment:
Does this also include screen capture sources?

Reply:
It doesn't, those are currently marked with a media class of  Stream/Output/Video